source/find_getmore_killcursors_commands.rst

```diff
@@ -365,18 +365,35 @@ More detailed information about the interaction of the **secondaryOk** with **OP
 Behavior of Limit, skip and batchSize
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The new **find** command has different semantics to the existing 3.0 and earlier **OP_QUERY** wire protocol message. The **limit** field is a hard limit on the total number of documents returned by the cursor no matter what **batchSize** is provided.
-
-Once the limit on the cursor has been reached the server will destroy the cursor and return a **cursorId** of **0** in the **OP_REPLY**. This differs from existing **OP_QUERY** behavior where there is no server side concept of limit and where the driver **MUST** keep track of the limit on the client side and **MUST** send a **OP_KILL_CURSORS** wire protocol message when it limit is reached.
-
-When setting the **batchSize** on the **find** and **getMore** command the value MUST be based on the cursor limit calculations specified in the `CRUD`_ specification. 
-
-In the following example the **limit** is set to **4** and the **batchSize** is set to **3** the following commands are executed.
+The new **find** command has different semantics to the existing 3.0 and earlier
+**OP_QUERY** wire protocol message. The **limit** field is a hard limit on the
+total number of documents returned by the cursor no matter what **batchSize** is
+provided. This includes other limiting operations, such as the **$limit**
+aggregation pipeline stage. This differs from existing **OP_QUERY** behavior
+where there is no server side concept of limit and where the driver **MUST**
+keep track of the limit on the client side and **MUST** send a
+**OP_KILL_CURSORS** wire protocol message when it limit is reached.
+
+When setting the **batchSize** on the **find** and **getMore** command the value
+MUST be based on the cursor limit calculations specified in the `CRUD`_
+specification.
+
+Note that with 5.0, the server-side handling of cursors with a limit has
+changed. Before 5.0, some cursors were automatically closed when the limit was
+reached (e.g. when running **find** with **limit**), and the reply document did
+not include a cursor ID. Starting with 5.0, all cursor-producing operations will
+return a cursor ID if the end of the batch being returned lines up with the
+limit on the cursor. In this case, drivers MUST send an additional **getMore**
```

Comment:
The way this is written, it seems like a special case that needs to be added.  But don't we want to remove the special case?  Should drivers just exhaust the cursor by executing getMore with the same batch size, up until the server returns a cursorId of 0?  And just let the server worry about enforcing the limit?

Reply:
The goal here is to not change the behaviour entirely (removing limit tracking), as that'd be more effort than fixing the tests. Since the cursor is open, either `killCursors` or `getMore` are expected. For any driver that doesn't track the limit, the presence of a cursor ID means that they'll do another `getMore` request, which is why I added this behaviour. I'm all for dropping this tracking entirely, especially since this behaviour doesn't cover cursors returned from `aggregate` (where we don't have an explicit limit), but would like to defer that to Q3 and DRIVERS-1448. For now I only want to ensure that the behaviour is documented and drivers don't leave any open cursors on servers.